presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -290,24 +294,27 @@ public synchronized void scheduleRequestIfNecessary()
             return;
         }
 
-        long neededBytes = bufferCapacity - bufferBytes;
-        if (neededBytes <= 0) {
-            return;
-        }
-
-        int clientCount = (int) ((1.0 * neededBytes / averageBytesPerRequest) * concurrentRequestMultiplier);
-        clientCount = Math.max(clientCount, 1);
-
-        int pendingClients = allClients.size() - queuedClients.size() - completedClients.size();
-        clientCount -= pendingClients;
-
-        for (int i = 0; i < clientCount; i++) {
+        long availableBytes = bufferCapacity - bufferBytes;
+        while (reservedBytes.get() < availableBytes) {
             HttpPageBufferClient client = queuedClients.poll();
             if (client == null) {
                 // no more clients available
                 return;
             }
-            client.scheduleRequest();
+            // reservedBytes cannot be greater than availableBytes as this is the only place to bump up reservedBytes
+            long maxBytes = Math.min(client.getTotalPendingSize().toBytes(), availableBytes - reservedBytes.get());
+            reservedBytes.addAndGet(maxBytes);
+            client.scheduleRequest(new DataSize(maxBytes, BYTE));
+        }
+
+        // drain all the clients that do not have a pending page
```

Comment:
Consider the following queue with each number denoting the pending pages (in bytes) for each client.
Suppose `availableBytes` is 8.

```
   -------------------
<- 0 0 1 2 5 6 0 0 5 0  <- enqueue
   -------------------
```

The previous loop will allow the first 5 clients to fetch pages till the quota reaches 8:


```
   -------------------
<- 6 0 0 5 0  <- enqueue
   -------------------
```

But indeed those clients without pending pages should also send requests to the server to ask if there is any page available.
So, what this loop does is to drain the queue to be:

```
   -------------------
<- 6 5  <- enqueue
   -------------------
```

Those clients are not going to fetch real data; so it is safe doing so.
Also, those requests are long poll to prevent adding them back immediately.


Will add a comment for this.